File: app/src/main/java/com/yc/emotion/home/base/ui/fragment/BaseFragment.java
package com.yc.emotion.home.base.ui.fragment;

import android.os.Bundle;
import android.os.Message;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.yc.emotion.home.utils.WeakHandler;

import org.jetbrains.annotations.NotNull;

/**
 * Created by mayn on 2019/4/23.
 */

public abstract class BaseFragment extends Fragment  {

    private String fragmentTag = "";
//    public String LOG_TAG_FRAGMENT_NAME;


//    protected WeakHandler mHandler ;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
//        LOG_TAG_FRAGMENT_NAME = getClass().getName() + " ";
        Log.d("ClassName", "onCreate: ------Fragment :" + getClass().getName());
//        mHandler= new WeakHandler(this);
        return onFragmentCreateView(inflater, container, savedInstanceState);
    }

    protected abstract View onFragmentCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState);

    public String getFragmentTag() {
        return fragmentTag;
    }

    public void setFragmentTag(String fragmentTag) {
        this.fragmentTag = fragmentTag;
    }



}

File: app/src/main/java/com/yc/emotion/home/base/ui/fragment/BaseFragment.kt
package com.yc.emotion.home.base.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.yc.emotion.home.base.domain.model.IModel
import com.yc.emotion.home.base.presenter.BasePresenter
import com.yc.emotion.home.base.view.IView

/**
 * Created by mayn on 2019/4/24.
 */
abstract class BaseFragment<P : BasePresenter<out IModel, out IView>?> : Fragment(), IView {

    protected lateinit var rootView: View
    private var isInitView = false
    private var visible: Boolean = false

    protected var mPresenter: P? = null

    internal var fragmentTag = ""


    //    protected BasePresenter mPresenter;
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(getLayoutId(), container, false)

        return rootView
    }

    protected open fun initBundle() {}
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initBundle()
        initViews()
        isInitView = true
        isCanLoadData()
        mPresenter?.subscribe()
    }


    override fun onResume() {
        super.onResume()
        if (!isHidden && isResumed) {
            visible = true
            isCanLoadData()
        }
    }

    override fun onPause() {
        super.onPause()
        visible = false
    }

//    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
//        super.setUserVisibleHint(isVisibleToUser)
//        //isVisibleToUser这个boolean值表示:该Fragment的UI 用户是否可见，获取该标志记录下来
//        if (isVisibleToUser) {
//            visible = true
//            isCanLoadData()
//        } else {
//            visible = false
//        }
//    }//防止重复加载数据

    //所以条件是view初始化完成并且对用户可见
    private fun isCanLoadData() {
        //所以条件是view初始化完成并且对用户可见
        if (isInitView && isVisible) {
            lazyLoad()

            //防止重复加载数据
            isInitView = false
//            visible = false
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        mPresenter?.unSubscribe()

    }

    /**
     * 加载要显示的数据
     */
    protected abstract fun lazyLoad()

}